### DISCUSSION TITLE
Nested Patch Updates?

### DISCUSSION THREAD

**Discussion by imaurer** [pydantic/pydantic#8615] [Question]
Looking for a "best practice" approach for doing "nested patch updates" using Pydantic.

For example,

```python
from pydantic import BaseModel

class B(BaseModel):
    pk: int
    b1: str
    b2: str

class A(BaseModel):
    a1: str
    a2: str
    bs: list[B]

original = A(a1="apple", a2="orange",
	   bs=[
	     B(pk=1, b1="banana", b2="carrot"),
	     B(pk=2, b1="tomato", b2="potato")
	   ])

partial_update = A(bs=[B(pk=2, b2="yam")])

# NOTE FOR SKIMMERS: THIS IS NOT REAL
# patch_update is not real, I am making this up for this question
updated = original.patch_update(partial_update)

expected = A(a1="apple", a2="orange",
             bs=[
                 B(pk=1, b1="banana", b2="carrot"),
                 B(pk=2, b1="tomato", b2="yam")
             ])

assert expected == updated
```

The challenge is that the construction of the `partial_update` object would fail due to the missing required fields (a1, a2, bs.b1). 

Right now, I am doing something similar to the notes in the following discussions and web pages linked below. Basically, I am creating a separate class with less stringent requirements and then leveraging a mix of model_dump, exclude_unset, model_copy, etc. to accomplish this goal.

Some tricky bits embedded with the idea of a "primary key" (pk field) for identifying and isolating child objects in a list.

This pattern is coming up quite a bit for me based on my work with LLMs and Pydantic.

Curious if there was a better way.

In summary, the 3 key features I am looking for are:
1. Autogenerated "Update" Model that isn't invalid due to missing requirements.
2. Primary Key (pk) capability for nested-object identification by partial update object
3. "patch_update" function that wraps the dump/unset, copy, etc. capabilities

It's OK if there isn't a better way and the logic should just live in my apps, I was just curious if it was something that could get baked into Pydantic proper to eliminate the work of keeping multiple versions of classes in-sync.

cheers and keep up the GREAT work,
Ian

P.S. Some links I found:

FastAPI Discussion & Documentation
https://github.com/tiangolo/fastapi/discussions/8258
https://fastapi.tiangolo.com/tutorial/body-updates/

Pydantic Discussion
https://github.com/pydantic/pydantic/issues/7387

P.P.S. I learned about JSON PATCH (RFC 6902) which I discarded as being ugly and losing the value of Pydantic. The "path" attribute is not ideal. However, I  thought it was worth sharing with this group since it would technically solve my problem since I know LLMs know about RFCs.

JSON Patch (Discarded, Ugly Format)
https://datatracker.ietf.org/doc/html/rfc6902
https://en.wikipedia.org/wiki/JSON_Patch
https://github.com/stefankoegl/python-json-patch
https://github.com/json-patch/json-patch-tests

**Comment by imaurer** [comment]
I found this library which covers a lot of my thoughts:
https://github.com/zmievsa/pydantic-duality

cc: @zmievsa

**Comment by samuelcolvin** [comment]
See my reply to [your tweet](https://twitter.com/imaurer/status/1759928150706855989) and his discussion](https://discuss.python.org/t/introduce-partial-for-typeddict/45176/32). In particular, [my answer here](https://discuss.python.org/t/introduce-partial-for-typeddict/45176/13).
